lib/Monads/trace/Trace_More_RG.thy

```diff
@@ -0,0 +1,685 @@
+   \<Longrightarrow> guar_cond G s0 tr"
+  unfolding validIE_def
+  by (frule(2) use_validI_guar, simp)
+
+lemma validI_preservation_ex:
+  assumes x: "\<And>x P. \<lbrace>\<lambda>s0 s. P (f s0 s x :: 'b)\<rbrace>,\<lbrace>R\<rbrace> m \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0 s. P (f s0 s x)\<rbrace>"
+  shows      "\<lbrace>\<lambda>s0 s. P (f s0 s :: 'a \<Rightarrow> 'b)\<rbrace>,\<lbrace>R\<rbrace> m \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0 s. P (f s0 s)\<rbrace>"
+  apply (clarsimp simp: validI_def validI_prefix_closed[OF x] guar_by_rg[OF x])
+  apply (erule subst[rotated, where P=P])
+  apply (rule ext)
+  apply (erule use_validI [OF _ x _refl])
+  apply simp
+  done
+
+lemma whenE_invI:
+  assumes a: "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>_. P\<rbrace>"
+  shows "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> whenE Q f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>_. P\<rbrace>"
+  by (wpsimp wp: a)
+
+lemma ifM_throwError_returnOkI:
+  "\<lbrace>Q\<rbrace>,\<lbrace>R\<rbrace> test \<lbrace>G\<rbrace>,\<lbrace>\<lambda>c s0 s. \<not> c \<longrightarrow> P s0 s\<rbrace>
+   \<Longrightarrow> \<lbrace>Q\<rbrace>,\<lbrace>R\<rbrace> ifM test (throwError e) (returnOk ()) \<lbrace>G\<rbrace>,\<lbrace>\<lambda>_. P\<rbrace>, -"
+  unfolding ifM_def
+  apply (fold liftE_bindE)
+  apply wpsimp
+   apply assumption
+  apply simp
+  done
+
+lemmas state_unchanged_rg = in_inv_by_rgD [THEN sym]
+lemmas last_state_unchanged_rg = last_st_in_inv_by_rgD [THEN sym]
+
+\<comment> \<open>FIXME MC: name? move (both this one and validI in More_VCG)\<close>
+lemma validI_I:
+  assumes px: "prefix_closed S"
+  assumes gc: "\<And>s0 s tr res. \<lbrakk> P s0 s; (tr, res) \<in> (rely S R s0 s)\<rbrakk> \<Longrightarrow> guar_cond G s0 tr"
+  assumes rl: "\<And>s0 s tr r s'. \<lbrakk> P s0 s; (tr, Result (r, s')) \<in> (rely S R s0 s) \<rbrakk> \<Longrightarrow> Q r (last_st_tr tr s0) s'"
+  shows "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> S \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>"
+  unfolding validI_def using px gc rl by safe
+
+lemma opt_return_pres_lift_rg:
+  assumes x: "\<And>v. \<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f v \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv. P\<rbrace>"
+  shows      "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> case x of None \<Rightarrow> return () | Some v \<Rightarrow> f v \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv. P\<rbrace>"
+  by (wpsimp wp: x)
+
+\<comment> \<open>FIXME MC: this would be a pain to prove for validI, can it be deleted?
+lemma validI_return_unit:
+  "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f >>= (\<lambda>_. return ()) \<lbrace>G\<rbrace>,\<lbrace>\<lambda>r. Q\<rbrace> \<Longrightarrow> \<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>r. Q\<rbrace>"
+  by (auto simp: validI_def in_bind in_return Ball_def)\<close>
```

Comment:
I think that can be removed. `wp` should be fine to deal with the bind/return pattern without a specific rule.

Reply:
Another to add to the list to try removing from nondet.